### PR TITLE
Add hasAux support to grad, valueAndGrad and vjp

### PR DESCRIPTION
**Pull request by jimfleming**
Adds `{ hasAux: true }` option to `vjp`, `grad`, and `valueAndGrad` for returning auxiliary outputs alongside gradients.

  ```ts
  // vjp returns [output, vjpFn, aux]
  const [y, vjpFn, aux] = vjp(f, { hasAux: true }, x);

  // grad returns [gradient, aux]
  const [gradient, aux] = grad(f, { hasAux: true })(x);

  // valueAndGrad returns [value, gradient, aux]
  const [value, gradient, aux] = valueAndGrad(f, { hasAux: true })(x);
```

### CODE REVIEW

**Comment by ekzhang** [comment]
I think we might be able to use TypeScript type overloads!

File: src/frontend/linearize.ts [hunk 1]
```diff
@@ -953,9 +954,63 @@ function vjpFlat(
...
+      const [cotangentsOutFlat, cotangentTree] = treeFlatten(cotangentsOut);
+      if (!mainTree.value!.equals(cotangentTree)) {
+        throw new TreeMismatchError("vjp", mainTree.value!, cotangentTree);
+      }
+      const cotangentsInFlat = fVjpFlat(...cotangentsOutFlat.map(pureArray));
+      return treeUnflatten(inTree, cotangentsInFlat);
+    }) as OwnedFunction<(cotangents: any) => any>;
+    fVjp.dispose = dispose;
+
+    return [primalsOut, fVjp, aux];
```
Comment (ekzhang) [inline src/frontend/linearize.ts]: Is there any way to reduce code duplication by merging this with the other code path except for the aux part?

File: src/frontend/linearize.ts [hunk 2]
```diff
@@ -953,9 +954,63 @@ function vjpFlat(
...
+
+  const firstArg = args[0];
+  if (firstArg && typeof firstArg === "object" && "hasAux" in firstArg) {
+    opts = firstArg;
+    primalsIn = args.slice(1);
+  }
+
   const [primalsInFlat, inTree] = treeFlatten(primalsIn);
+
+  if (opts?.hasAux) {
```
Comment (ekzhang) [inline src/frontend/linearize.ts]: Hmm yeah, this is tricky! If a function f is `X => [Y, Aux]`, perhaps we can handle vjp aux by transforming it to a non-aux vjp

```ts
function vjpWithAux(f, ...x) {
  let auxValue;
  const [y, fVjp] = vjp((...x) => {
    const [y, aux] = f(...x);
    auxValue = aux;
    return y;
  }, ...x);
  return [y, fVjp, auxValue];
}
```

I am really not sure though, this is surprisingly tricky

**Review by ekzhang — COMMENTED** [review]
Nice typed APIs! Looks great. I think there might be an edge case in how aux works — you're assuming the auxiliary data is arrays, but it can actually be any arbitrary object in the original JAX python library, it's actually not transformed by the system at all.

The code duplication is minor but I also feel something is slightly wrong with this implementation that I'm not sure of yet, maybe I need to think more about this. I wonder if the semantics in my comment below are correct or if they're wrong.

**Comment by ekzhang** [comment]
In particular, this works! Hmm…

```
>>> jax.grad(lambda x: (jnp.sin(x), open('.bashrc', 'r')), has_aux=True)(2.0)
(Array(-0.41614684, dtype=float32, weak_type=True), <_io.TextIOWrapper name='.bashrc' mode='r' encoding='UTF-8'>)
```

But this also works:

```
>>> jax.grad(lambda x: (jnp.sin(x), (x, x)), has_aux=True)(2.0)
(Array(-0.41614684, dtype=float32, weak_type=True), (TypedFloat(2.0, dtype=float32), TypedFloat(2.0, dtype=float32)))
```

So maybe the way jax handles aux is by lowering as many tracers as they can, but also not recursing into arbitrary data? Very tricky.

**Comment by ekzhang** [comment]
Thanks for the detail, I think I was feeling a bit out of my depth in reviewing this haha.

I think that makes sense about the auxiliary values being a pytree of arrays, now that I think about it more. Otherwise it wouldn't really be that useful — I think, still haven't fully wrapped my head around it 😅. In any case, you have a good argument for why it's useful with pytrees of arrays

Down to merge now or later if you think we can avoid as much duplication between the aux/non-aux code paths!

**Comment by ekzhang** [comment]
Thanks! I realize now when reading this that it doesn't handle `{ hasAux: false }` as a case. Sorry this is my bad though, I should've caught this on the first review rather than doing a back-and-forth.

If it's alright I'm just going to make a breaking library change here and make `vjp()` and `linearize()` take in arrays of inputs (consistent with `jvp()`) so we don't need to worry too much, then edit your PR to work with the new setup! I think it's ok since they aren't the most common primitives and it'll make the runtime interface cleaner.

**Comment by ekzhang** [comment]
Alright that was quite complex!! Thanks for the great test suite by the way, it made it easy to test the interaction of this with `argnums` as well.
